lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -600,6 +600,37 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
   return tryToConcatenateStrings(AI, Builder);
 }
 
+/// Determine the pattern for global_addr..
+/// %3 = global_addr @$P : $*SomeP
+/// %4 = init_existential_addr %3 : $*SomeP, $SomeC
+/// %5 = alloc_ref $SomeC
+/// store %5 to %4 : $*SomeC
+static SILValue findInitExistentialFromGlobalAddr(GlobalAddrInst *GAI,
+                  CopyAddrInst *CAI) {
+  /// Check for a single InitExistential.
+  InitExistentialAddrInst *SingleIE = nullptr;
+  auto *BB = CAI->getParent();
+  for (auto Use : GAI->getUses()) {
+    auto *User = Use->getUser();
+    if (auto *InitExistential = dyn_cast<InitExistentialAddrInst>(User)) {
+      /// Simple dominance check: both InitExistential and CAI are in 
+      /// the same basic block and not more than one InitExistential 
+      /// occurs before CAI.
+      auto *IEBB = InitExistential->getParent();
+      if (BB == IEBB &&
```

Comment:
You are ignoring ``init_existential_addr`` that is in a different block. In such a case you should bail, no? So this should be:

```
if (BB != IEBB) {
  return SILValue();
}

if (std::find_if(...) == BB->end()) {
  return SILValue();
}

if (SingleIE) {
  return SILValue();
}

SingleIE = InitExistentialAddr;
```

Assuming that I am correct, can you add a test case for this as well?

Reply:
The first condition you mentioned for BB check is not correct (premature return)! I need one occurrence of IE in BB -- IEs in other BB's are skipped  -- Ideally I just want to iterate on the uses of GAI in BB only -- is there any API for this? Current code is fine, but does little more work -- iterates over all uses doing nothing.

Reply:
I think what I wrote is correct depending on what we are trying to implement ; ). That is: to be very conservative, I was trying to make it so that we bailed if we saw /any/ other init_existential_addr in other BBs on the GAI. In that case we should bail/fail early in the way that I mentioned. That being said, I do see your point that such things should not matter if the init_existential_addr and the copy_addr are in the same block and there is a single init_existential_addr in the block before the copy_addr.

In terms of the API, I am not sure if one exists. That being said, I wonder if the better way to do this is to put all of the uses of GAI into a set and then just iterate backwards from the CAI to the beginning of the block looking for the target. Specifically:

```
llvm::SmallPtrSet<InitExistentialAddrInst *, 8> Initializers;
for (auto *Use : GAI->getUsers()) {
  if (auto *I = dyn_cast<InitExistentialAddrInst>(Use->getUser()) {
    Initializers.insert(I);
  }
}

if (Initializers.empty())
  return SILValue();

SILValue Result;
for (auto II = CAI->getIterator().getReverse(), IE = CAI->getParent()->rend(); II != IE; ++II) {
    if (!Initializers.count(II)) {
      continue;
    }

    if (Result)
      return SILValue();

   Result = *II;
}
```